### PR TITLE
Virgo soc driver atcdmac100

### PR DESCRIPTION
**Pull request by junaidaslamRS**
- DMA IP works ok as Memory mapped reading of external flash via DMA produced successful results as shown in the image below. GPT Also Working fine.
image

### CODE REVIEW

File: dts/bindings/dma/andestech,atcdmac100.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+               0x4: Quad word (128 bits)
+               0x5: Eight word (256 bits)
+               0x6-0x7: reserved
+        -bit 10 : Priority level
+               0x0: lower priority
+               0x1: higher priority
+
+    examples for andes_v5_ae350 DMA instance
+     dma0: dma0@f0c00000 {
+         compatible = "andestech,atcdmac300";
```
Comment (rahul-r-shah-pub) [inline dts/bindings/dma/andestech,atcdmac100.yaml]:
```suggestion
         compatible = "andestech,atcdmac100";
```

File: dts/riscv/rapidsilicon/rapidsi_virgo.dtsi [hunk 2]
```diff
@@ -162,12 +162,12 @@
...
  
 		 dma0: dma@A0600000 {
-			 compatible = "andestech,atcdmac300";
+			 compatible = "andestech,atcdmac100";
 			 reg = <0xA0600000 0xFFFF>;
 			 interrupts = <27 9>;
 			 interrupt-parent = <&eclic>;
 			 dma-channels = <8>;
-			 dma-requests = <16>;
+			 dma-requests = <32>;
```
Comment (rahul-r-shah-pub) [inline dts/riscv/rapidsilicon/rapidsi_virgo.dtsi:170]: Perhaps dma-channel-mask should also be set. This is the bitmask of channels that are not used up by firmware and available to kernel. It is possible that kernel may want to use some of the channels. I dont know what the default bit mask is but all 1s or all 0s both would be wrong. Currently we are using 0 and 1 for SPI, but we may end up using more for Configuration and everytime we use a channel, we will have to change this. Another question is that can we get an available channel from the kernel (that means make all the channels available to the kernel) when we want to use in FW?

File: dts/riscv/rapidsilicon/rapidsi_virgo.dtsi [hunk 3]
```diff
@@ -150,7 +150,7 @@
 		 spi0: spi@A0800000 {
 			 compatible = "andestech,atcspi200";
```
Comment (rahul-r-shah-pub) [inline dts/riscv/rapidsilicon/rapidsi_virgo.dtsi:151]: This file name in the code hierarchy is andestech.atcspi200.yaml, it should be andestech,atcspi200.yaml. Notice the full-stop should be replaced by comma

File: dts/riscv/rapidsilicon/rapidsi_virgo.dtsi [hunk 4]
```diff
@@ -150,7 +150,7 @@
...
 			 reg = <0xA0800000 0xFFFF>;
-			 interrupts = <24 5>;
+			 interrupts = <24 10>;
```
Comment (rahul-r-shah-pub) [inline dts/riscv/rapidsilicon/rapidsi_virgo.dtsi:153]: My only concern is that WDT is considered as -1 interrupt. interrupt to the RISC-V code 0-18 are internal, the WDT is -1 and then timer onward is 19 onward. this way SPI is 24 interrupt number. I want to figure out what the WDT is connected to? Is it connected to reset and is the Timer connected to 19?

**Review by rahul-r-shah-pub — CHANGES_REQUESTED** [review]
Some changes and some things to discuss

File: dts/riscv/rapidsilicon/rapidsi_virgo.dtsi (hunk 4 above)
Comment (junaidaslamRS) [reply dts/riscv/rapidsilicon/rapidsi_virgo.dtsi:153]: Yes, the PIT is connected to interrupt 19 as it does trigger. I have also been thinking about the same WDT -1 interrupt question and how in Zephyr this gets connected. I still have to dive into WDT to check several things.

File: dts/riscv/rapidsilicon/rapidsi_virgo.dtsi (hunk 2 above)
Comment (junaidaslamRS) [reply dts/riscv/rapidsilicon/rapidsi_virgo.dtsi:170]: As per Gleb, any peripheral can use any DMA channel. There is no hard restriction to it. I have not come across a scenario where Kernel could use DMA for its internal operations.

Comment (rahul-r-shah) [reply dts/riscv/rapidsilicon/rapidsi_virgo.dtsi:170]: Yes, I agree there are no restrictions from the HW PoV, but do we in the firmware code use specific channels? If so they should not be available to the kernel OR we could let everything be available to the kernel and figure out a kernel API / top level DMA driver API, which gives us the free DMA channels and in this case the kernel should maintain the taken up and free DMA channels. This is my understanding based on logic and I have not really gone deep into the Zephyr DMA setup. Anyways, this is fine for now, but you may want to check it out

File: samples/hello_world/src/main.c [hunk 5]
```diff
@@ -112,29 +289,43 @@ int main(void)
...
-	if((flash == NULL) || (!device_is_ready(flash))) {
+	if(spi == NULL) {
+		printf("%s spi has status disabled...%s\n", ATTR_ERR,ATTR_RST);
+	} else {			
+		printf("%s spi Object is Created. Test Via DMA\n", ATTR_INF);				
+		if(flash == NULL) {
+			printf("%s flash has status disabled...%s\n", ATTR_ERR,ATTR_RST);
+		} else {
+			printf("%s flash Object is Created. Initialize to Test Via DMA%s\n", ATTR_INF,ATTR_RST);						
+			if(flash != NULL) {
```
Comment (rahul-r-shah) [inline samples/hello_world/src/main.c]: Do we need this extra check? Youve already check this above flash == NULL, or do you mean to be checking for DMA here?

File: drivers/dma/dma_andes_atcdmac100.c [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+#define DMA_INT_STATUS_TC_VAL(x)	FIELD_GET(DMA_INT_STATUS_TC_MASK, (x))
+#define DMA_INT_STATUS_ABORT_VAL(x)	FIELD_GET(DMA_INT_STATUS_ABORT_MASK, (x))
+#define DMA_INT_STATUS_ERROR_VAL(x)	FIELD_GET(DMA_INT_STATUS_ERROR_MASK, (x))
+#define DMA_INT_STATUS_CH_MSK(ch)	(0x111 << ch)
+
+typedef void (*atcdmac100_cfg_func_t)(void);
+
+struct __attribute__((packed, aligned(4))) chain_block {
+	uint32_t ctrl;
+	uint32_t srcaddrl;
```
Comment (rahul-r-shah) [inline drivers/dma/dma_andes_atcdmac100.c]: Hey, whats the 'l' at the end of srcaddr?

File: samples/hello_world/src/main.c [hunk 7]
```diff
@@ -47,9 +50,9 @@ void Flash_Test(const struct device *flash, uint32_t FLASH_ADDR, uint8_t EVEN_OD
...
 			printf("%s\nFlash erase at 0x%08x did not produce correct results%s\n", ATTR_ERR, FLASH_ADDR,ATTR_RST);
-			for(uint8_t i = 0; i < FLASH_RW_SIZE; i++) {
-				printf("%s 0x%02x%s", ATTR_INF, flash_data_read[i],i%FORMATTER==0?"\n":"");			
-			} printf("\n");
+			// for(uint8_t i = 0; i < FLASH_RW_SIZE; i++) {
+			// 	printf("%s 0x%02x%s", ATTR_INF, flash_data_read[i],i%FORMATTER==0?"\n":"");			
+			// } printf("\n");
 		} else {
 			printf("%s\nSuccessfully performed erase to flash with code:%d%s\n", ATTR_INF, errorcode,ATTR_RST);
 		}
```
Comment (rahul-r-shah) [inline samples/hello_world/src/main.c:58]: LInes 51 to 58 do not seem to make sense. They have already been covered in the if errorcode < 0 in line 41 above

File: samples/hello_world/src/main.c [hunk 8]
```diff
@@ -5,17 +5,20 @@
  */
 
 #include <zephyr/kernel.h>
+#include <zephyr/drivers/counter.h>
```
Comment (rahul-r-shah) [inline samples/hello_world/src/main.c:8]: Reiterating that we should at some point in time revert to the original hello_world main code and have our code in the boards maybe and compile and run from there

**Review by rahul-r-shah — CHANGES_REQUESTED** [review]
Last of reviews done, couple of changes

File: drivers/dma/dma_andes_atcdmac100.c (hunk 6 above)
Comment (junaidaslamRS) [reply drivers/dma/dma_andes_atcdmac100.c]: in ATCDMAC300 we have low and high registers. I removed all the srcaddrh type variables and left srcaddrl type. May be i can rename it to remove this l part as well.

File: samples/hello_world/src/main.c (hunk 5 above)
Comment (junaidaslamRS) [reply samples/hello_world/src/main.c]: Actually now we dont need it now. Earlier i was doing deferred initialization of flash. There it was needed. that extra initialization line is removed. I will remove this.

File: samples/hello_world/src/main.c (hunk 7 above)
Comment (junaidaslamRS) [reply samples/hello_world/src/main.c:58]: This is actually serving a different purpose. After erasing the flash, if we don't see all 0xFF values, then it means the flash erase did not go well. The error check at line 41 suggests an overall protocol error. This is where I had found earlier that the erase command supported by our flash was not part of the DTS SFDP Header bytes. Then I modified that header to make it correct.

**Review by rahul-r-shah — APPROVED** [review]
LGTM

**Pull request merged**